### PR TITLE
Move 'dependencies.gradle' to buildSrc as 'libraries.kt'

### PR DESCRIPTION
**Pull request by jjohannes**
1. convert from groovy to kotlin-dsl
2. move to buildSrc so that we have statically typed values

There is some discussion around if we want to do it like this and if this is the way that we would recommend to others: https://github.com/gradle/dotorg-docs/issues/192

I would be interested in your thoughts @eskatos @bamboo @bigdaz @melix @ljacomet.

If we think we need to do something else, but this is a step in the right direction, we can merge this PR and continue to refine it afterwards.

### CODE REVIEW

**Comment by big-guy** [comment]
The big downside to doing this is that any change in dependencies will cause all tasks defined in `buildSrc` to be invalidated/rerun.

**Comment by jjohannes** [comment]
Despite my scepticism to this solution, I started to convert more subproject scripts to kotlin and fix the dependency declarations in these scripts in the process (api/implementation separation). I did this based on this PR (I can change it if we do not want to merge this one) - #4498.

One thing to fix here is that libraries (gson, asm, ...) should be imported direct by the projects that need them. Since they do not provide types to the API. They just happen to be used in multiple projects. But not in all projects in the hierarchy. So only selected projects should have an `implementation` dependency on them.

Then you have this situation, where you import the same library several times in different places. And here it is really **very convenient** to define them in one place and then have the IDE support to refer them wherever you need them. I could also imagine a nice convenient plugin that contains constants for all widely-used dependencies on jcenter...

**Comment by ljacomet** [comment]
What feels weird to me here is that we end up having a custom type to declare the dependencies to use, that then gets converted in the right type by usage.

If we go for a type safe definition, cannot we use something that Gradle will be able to natively interpret, so that having the following definition (using the syntax / types from this PR) `lib("junit:junit", "4.12", "not yet ready for JUnit 5")` and using it as `implementation Libraries.junit` is equivalent to having:
```
implementation "junit:junit:4.12" {
    because "not yet ready for JUnit 5"
}
```

In short, if I have a rich type containing `group`, `name`, `version` and `because`, why should I decompose it again?

**Comment by big-guy** [comment]
Can we inject a `DependencyHandler` and use `Dependency`s directly (instead of referring to the coordinates explicitly)?

**Comment by melix** [comment]
I for one don't like the precedent that this creates. `buildSrc` is for build logic. Not configuration. The libraries we use are configuration, and therefore should be declared in a `libraries.gradle(.kts)` or `libraries.properties` or `libraries.json` file in the build itself. The fact that accidentally this enables IDE support is wrong reasoning to me. It's convenient, but we should find a way to have completion, or nice support, without migrating everything to buildSrc. Let's not add confusion around what `buildSrc` is for.

**Comment by bamboo** [comment]
> buildSrc is for build logic. Not configuration.

`buildSrc` is for sharing definitions across all build scripts in a build. What roles those definitions play is at the discretion of the user.

I think it's more useful to think in terms of expected rate of change. Things that change often are usually better managed outside `buildSrc` because of the overhead imposed by the changes to the classpath.

Back to the issue at hand, a good compromise here could be to declare a Gradle project extension, say, `Libraries` defined in `buildSrc` which gets populated from one build script and consumed from the rest.

Here's a sketch:

```kotlin
// buildSrc/**/Libraries.kt
class Libraries {
  lateinit var junit: Library
  ...
}

// buildSrc/**/LibrariesPlugin.kt
    project.extensions.create<Libraries>("libraries")

// main build script
plugins {
...
  id("libraries")
...
}
...
libraries {
  junit = Library("....")
  ...
}

// consumer build script
dependencies {
   implementation(libraries.junit)
}
```

**Comment by eskatos** [comment]
The sketch suggested by @bamboo above could also be enhanced following @ljacomet suggestion, by using `Dependency` subtypes directly. Gradle extensions can get to `DependencyHandler`.

**Comment by melix** [comment]
I mean no offense here, but I think this is all cock-eyed. It's not because we have a static language that we _need_ everything static with everything referencing everything everywhere. Where is the simplicity? Can we seriously propose to a new user to create a file in `buildSrc/` that _pre-declares_ a variable name, then initialize it in a build script? I cannot imagine _a second_ having to explain that to beginners, or even existing Gradle users. I don't want Gradle to become Sbt. I want to keep the simplicity of build scripts, and the simplicity is that I want to declare a dependency like this:

```
dependencies {
    implementation 'com.foo:bar:1.0'
}
```

Do Maven users want completion like this? No, but what they *do* have, instead, and which is IMO far more interesting for 90% of our users, is completion of the group id, artifact id and version of dependencies as they type it. This is the kind of content assist _most_ users are looking for. So while we _could_ declare dependencies like this, half split in `Libraries.kt`, half in a build script, and worse, think that we can start putting _configuration_ in `buildSrc` is turning Gradle into a different build tool, which, I'm afraid, is not the one I want.

If we go that route, the idiomatic build we're going to show will be:

`buildSrc/..../lots of fancy Kotlin scripts`
`build.gradle.kts: apply { plugin 'configure-my-build' }`

This *isn't* right to me. That's not because the only way to get content assist today is working with `buildSrc` that we have to endorse it. There MUST be a clear separation of contents, and build scripts must remain _readable_, especially when read on GitHub or in diffs. So the goal should be, IMO:

```
dependencies {
    implementation 'com.foo:<CTLR+SPACE> // <-- show me some nice completions
}
```

and for the advanced use case like us, where you want to define the libraries you use separately _for reusability_, it must be something like this:

```
apply {
   best-practice-dependencies-plugin
}

dependencies {
    implementation libraries.foo<CTLR+SPACE> // <-- autocompletes and maybe replaces it with libraries['com.foo']
}
```

And if it means having a convention that if you apply the "best practice plugin" then you must have a file named `libraries.json` with "library definitions", or a Kotlin file, or a Groovy one, ... then the IDE _can_ support it, that's what IDEs do, they support external tools and support their features.

I think that searching for type safety for the sake of type safety is insane. We must aim for usability. What most people want is completion and our sole principle should be to avoid imaginative complex solutions when there are nice existing ones that do not necessitate special IDE support.

That's precisely why I said previously that I didn't want to come with a pattern, or best practice at the moment for the dependency management side of things, because we're introducing lots of new features, and that we may create a "best practices plugin" out of it. But we must really think about most of our user base, when they would read this:

To add a dependency on junit, do this:

1. open `buildSrc/Libraries.kt` and add an entry `lateinit val junit: Library`
2. open your build script and add, if it's not there, the libraries extension in scope (`project.extensions.create("libraries", Libraries::class.java`)
3. add a library for junit in the `libraries` section `junit = Library("....")`
4. and now you can use it as a dependency in your dependencies block (`implementation (libraries.junit)`)

And we're not even talking about adding constraints, version preferences or causes for dependencies!

No, I really think what most people want is just a file which lists all the libraries they use with their version, and they want that to be autocompleted. Or they want the entry to be added automatically if they declare a dependency on something which is not yet there in the `dependencies` block. Especially on this topic, there are tons of folks that just don't understand that we cannot have a single flat list of "dependencies" (not even talking about scopes/configurations). Let's just not forget about them.

**Comment by bamboo** [comment]
> Can we seriously propose to a new user to create a file in buildSrc/ that pre-declares a variable name, then initialize it in a build script?

I understood we were talking about how to re-organize our existing build logic to take advantage of the goodness static typing can bring.

**Comment by melix** [comment]
Note that I'm saying this because I'm also sometimes tempted, when playing with Kotlin, to introduce an inline function here, an extension there, because _oh look how nice it is then_, but realize that what I'm doing is no longer trying to build my app, like 99% of our users would do, I'm just taking pleasure learning a new language.

**Comment by bamboo** [comment]
> I could also imagine a nice convenient plugin that contains constants for all widely-used dependencies on jcenter...

We spiked something like that for the Plugin Portal some time ago: https://github.com/gradle/kotlin-dsl/issues/185#issuecomment-265726472

**Comment by hansd** [comment]
I think it is very important that we agree on what our own build should demonstrate and provide to our engineers. I think our build should demonstrate how a complex build can be in a well-modeled end state, is straightforward to understand and is easy to use by developers. It should be a reference project for complex enterprise builds. That is where we need to provide more guidance. We want to provide helpful patterns for people dealing with complex build problems. People who write their first hello world Gradle build are much better served with our extensive samples and introductory documentation. 

The Gradle build serves first and foremost the needs of its users. So when you say `It is super nice to show that you can autocomplete dependencies and have completion`. I would agree. But much more important is that it is super nice to *have* completion of our dependencies.

I can not see much difference in terms of comprehensibility between putting configuration into scripts in the gradle/gradle directory or at a certain location in buildSrc. The big difference is that in one case it requires a complete recompile of the buildSrc code base, which is a cost we don't want to incur for things that frequently change.

I don't have a strong opinion on how we want to model our default dependencies except that we definitely want to have code completion support for them.

**Comment by jjohannes** [comment]
One solution I discussed with @ljacomet two weeks ago is to only define the library coordinates in `buildSrc`. Then it's like a convenient collection of names you use for auto-complete.

But the versions and reasons are defined in the build scripts (probably `distributions-dependencies.gradle.kts`).

Can we agree on this as a compromise for the next step (does not need to be the final solution)?

I can do that, but I will only do it if we also gonna merge it. So much files were replaced and moved by now, that I basically have to start this over.

**Comment by big-guy** [comment]
@jjohannes the native team should be taking up some of the build improvements this week.  I would hold off and we can come up with the next step.

**Comment by jjohannes** [comment]
I am closing this PR. We will have to redo this once we decided how to move forward.
